Network: add support for 'dangling' filter

KarthikNayak opened the pull request:
Closes #30825

Like its counterpart in images and volumes, introduce the dangling
filter while listing networks. When the filter value is set to true,
only networks which aren't attached to containers and aren't builtin
networks are shown. When set to false, all builtin networks and
networks which are attached to containers are shown.

Signed-off-by: Karthik Nayak <Karthik.188@gmail.com>

---

**- How I did it**

Mostly by mimicking the existing code, and writing new code for supporting this feature. 

**- How to verify it**

Either by running the unit test included, or by simply building and running
`docker network ls --filter dangling=true` and `docker network ls --filter dangling=false`

**- Description for the changelog**

Network: add support for 'dangling' filter

**- A picture of a cute animal (not mandatory but encouraged)**

<a data-flickr-embed="true"  href="https://www.flickr.com/photos/stys19850129/32571697242/in/photolist-R6Fcn3-RQjeYY-QXDLYx-RLUhEq-RCfsNW-Si72qN-RKjMKd-Sheb8e-RUY1kf-dDwovw-SsJbxz-oKG6TJ-RTRjBU-Sa2HfK-qbac9i-RJycyQ-RRwvDo-dVup61-Srj1YH-RZUnrk-QC16S4-QN1VQP-QPKwbd-RXjkut-r1iML3-QMaLtG-RM4oCQ-RTRjQu-RMrZqA-RRGYww-4irz6w-QZqTQx-qVN3pe-mJ6aTa-RMrZtw-Srj2N8-RZ2v5Z-S6KY4Q-nqm9TN-RhuARE-Sdy74x-7EKwGg-RQkBPJ-S5n1rX-oHRcJa-QyxPn1-R5fBh3-SfoWa6-RSGKFg-ReHBMn/" title="cat"><img src="https://c1.staticflickr.com/1/533/32571697242_78d0f82020.jpg" width="500" height="333" alt="cat"></a>
Credits: Flickr/NEKOFighter

ripcurld0 commented on api/server/router/network/filter.go:
```diff
@@ -29,6 +29,29 @@ func filterNetworkByType(nws []types.NetworkResource, netType string) ([]types.N
+			}
+		}
+
+	default:
```
Not sure whether this part is needed.
`danglingOnly` is bool, which means it can be true or false.

KarthikNayak replied:
Makes sense, will remove that :+1:

ripcurld0 commented on api/server/router/network/filter.go:
```diff
@@ -29,6 +29,29 @@ func filterNetworkByType(nws []types.NetworkResource, netType string) ([]types.N
+func filterNetworkByUse(nws []types.NetworkResource, danglingOnly bool) ([]types.NetworkResource, error) {
+	retNws := []types.NetworkResource{}
+
+	switch danglingOnly {
```
**NOTE**: This is a suggestion.
I would prefer get rid of the `switch` block and do the following:
```go
for _, nw := range nws {
    if !runconfig.IsPreDefinedNetwork(nw.Name) && len(nw.Containers) == 0 && danglingOnly {
        retNws = append(retNws, nw)
    }
}
```
I think it'll work.

KarthikNayak replied:
But how will this work when `dangling=false`. In that case we'd have to append non-dangling networks to retNws. Which is done only when, either, the network has been attached to more than a container or is a pre-defined network.

ripcurld0 replied:
Yeah, I knew I missed something.
What about setting a filter function instead of using a condition.
```go
filterFunc := func(nw types.NetworkResource) bool {
    if danglingOnly {
        return !runconfig.IsPreDefinedNetwork(nw.Name) && len(nw.Containers) == 0
    }
    return runconfig.IsPreDefinedNetwork(nw.Name) || len(nw.Containers) > 0
}
...
for _, nw := range nws {
    if filterFunc(nw) {
        retNws = append(retNws, nw)
    }
}
```

KarthikNayak replied:
That works too, but I don't see why we need to do this?

I mean the code currently tries to be consistent with the existing function filterNetworksByType, and staying consistent would mean better readability. Not that I'm against your suggestion, I just don't see any advantage in doing so.

vdemeester replied:
i like @ripcurld0 idea 😊

KarthikNayak replied:
@vdemeester I'm ok with changing to that, but that wouldn't be consistent with the existing code, that's all.

AkihiroSuda commented on api/server/router/network/filter.go:
```diff
@@ -29,6 +29,29 @@ func filterNetworkByType(nws []types.NetworkResource, netType string) ([]types.N
+	switch danglingOnly {
+	case true:
+		for _, nw := range nws {
+			if len(nw.Containers) == 0 && !runconfig.IsPreDefinedNetwork(nw.Name) {
```
This does not work for swarm networks, because `nw.Containers` does not contain containers running on other nodes

KarthikNayak replied:
I see, any suggestions how to go about this?

AkihiroSuda replied:
My opinion is to add `DryRun` to `RemoveNetworkRequest`.
https://github.com/docker/swarmkit/blob/450df1ca02727e79b4c58e8448173150f3c56c25/api/control.proto#L289-L292

cc @aluzzardi @aaronlehmann

KarthikNayak replied:
That would mean dropping this commit and re-working, i.e. adding the DryRun option to RemoveNetworkRequest and perhaps then have a this filter just call that?

AkihiroSuda replied:
yes, unfortunately..

aaronlehmann replied:
@AkihiroSuda: I'm not sure I understand correctly, but wouldn't be more straightforward to add `network` selectors to `ListTasksRequest` and `ListServicesRequest`? "List the tasks and services that depend on this network" feels better from an API perspective than "tell me if it's possible to delete this network".

AkihiroSuda replied:
@aaronlehmann Since even non-service containers can be attached to swarm networks, I think what we need is https://github.com/docker/docker/issues/24186 rather than `ListTaskRequest`.
But implementing #24186 seems hard (https://github.com/docker/docker/issues/24186#issuecomment-229744543)

cc @dperny

aaronlehmann replied:
Sorry, I'm not sure I follow. #24186 says "This is more of an enhancement, but I think in the context of a swarm/cluster, I may want to list the containers participating in the entire network, not just on the current node." Isn't the same thing as adding a `network` selector to `ListTasksRequest`?

It's actually not very difficult to support this, because tasks and services are already indexed by network in swarmkit's data store. #24186 is an old issue from before we were able to do this.

AkihiroSuda replied:
IIUC, `ListTasksRequest` does not work for `docker run` containers.
i.e.
```
$ docker network create --driver overlay --attachable foo
$ docker run --network foo ...
```

aaronlehmann replied:
I would expect it to work for `docker run` containers attached to networks. Those are represented by tasks, and I don't know of anything that filters them out in swarmkit (maybe there is something filtering them in the engine?).

KarthikNayak replied:
So basically this PR would be dependent on #24186, right?

AkihiroSuda replied:
Probably yes. But it would be OK if ListTasksRequest work.

KarthikNayak replied:
Now that https://github.com/docker/docker/pull/31710 is accepted, I'll finish this soon :+1:

allencloud commented:
I think this PR changes filters for ` GET /networks` API in the daemon side, so we should change swagger.yml.
In addition, docker_network_ls.md should be modified as well. @KarthikNayak

KarthikNayak commented:
@AkihiroSuda @aaronlehmann Do have a second look.

aaronlehmann commented on api/server/router/network/filter.go:
```diff
@@ -8,6 +8,26 @@ import (
+		}
+	case false:
+		for _, nw := range nws {
+			if len(nw.Containers) > 0 || len(nw.Services) > 0 || runconfig.IsPreDefinedNetwork(nw.Name) {
```
I believe these fields are provided by the gossip layer and are eventually consistent. Is it a problem if the dangling filter may provide results that are slightly out of date?

cpuguy83 replied:
I don't think it should be a problem as long *as the consumer knows* this is potentially racey.
Trying to rm a network from this list wouldn't really pose a problem except for potentially racey behavior from the state propagation, but this problem would exist in any case.

AkihiroSuda commented:
How does it work with `docker run` containers on other nodes?
https://github.com/docker/docker/pull/31710 does not seem to support them

KarthikNayak commented:
@AkihiroSuda It won't, since this is based off that patch

KarthikNayak commented on api/server/router/network/filter.go:
```diff
@@ -8,6 +8,26 @@ import (
+		}
+	case false:
+		for _, nw := range nws {
+			if len(nw.Containers) > 0 || len(nw.Services) > 0 || runconfig.IsPreDefinedNetwork(nw.Name) {
```
I don't think I know enough to answer that. I'll look around.

vdemeester replied:
cc @mavenugo @aboch

AkihiroSuda commented:
Then please add a doc so as to clarify that `dangling` network is not guaranteed to be dangling?

KarthikNayak commented:
I could do that, would that be ok though? @AkihiroSuda

AkihiroSuda commented:
@KarthikNayak 

I think it is ok 

cc @sanimej @aaronlehmann @thaJeztah

thaJeztah commented:
Yes, the situation with (e.g.) volumes is no different; a volume can be "dangling" when filtering, but there can be a race where it's being used by a container directly after that.

Moving this to code review

thaJeztah commented:
looks like it needs a rebase @KarthikNayak 😢

KarthikNayak commented:
Already on it :)

allencloud commented on api/swagger.yaml:
```diff
@@ -6277,6 +6277,7 @@ paths:
             - `name=<network-name>` Matches all or part of a network name.
             - `scope=["swarm"|"global"|"local"]` Filters networks by scope (`swarm`, `global`, or `local`).
             - `type=["custom"|"builtin"]` Filters networks by type. The `custom` keyword returns all user-defined networks.
+            - `dangling=["true"|"false"] Print only dangling networks`
```
add "0" and "1" ?
WDYT

KarthikNayak replied:
will do s/Print/Prints. 

About '0' and '1', I'm a little confused.

```
~ docker image ls --filter dangling=0
Error response from daemon: Invalid filter 'dangling=[0]'

~ docker volume ls --filter dangling=0
DRIVER              VOLUME NAME
```
So I'm on the fence with that

allencloud replied:
Yeah, I am just referenced to https://github.com/docker/docker/blob/master/api/swagger.yaml#L5089-L5091 for image prune dangling:

```
- `dangling=<boolean>` When set to `true` (or `1`), prune only
               unused *and* untagged images. When set to `false`
               (or `0`), all unused images are pruned.
```

vdemeester commented:
ping @mavenugo @fcrisciani @thaJeztah @vieux what is the status with this one ? 👼

vieux commented:
@vdemeester I'm still 👍 for this feature

thaJeztah commented:
I'm still 👍 as well, but it needs a rebase and the docs, completion-scripts need to be moved to https://github.com/docker/cli @KarthikNayak

cpuguy83 commented:
I went ahead and rebased this, updated based on code comments.

thaJeztah commented on api/server/router/network/filter.go:
```diff
@@ -73,6 +94,28 @@ func filterNetworks(nws []types.NetworkResource, filter filters.Args) ([]types.N
+		if fErr != nil {
+			return nil, fErr
+		}
+		danglingNet, err := filterNetworkByUse(displayNet, danglingOnly)
```
Looks like `filterNetworkByUse()` never returns an error, so we may just as well remove error handling here (and remove the `error` return from `filterNetworkByUse()`

KarthikNayak replied:
This is done! 
```suggestion
		danglingNet, err := filterNetworkByUse(displayNet, danglingOnly)
```

thaJeztah commented on api/server/router/network/filter.go:
```diff
@@ -73,6 +94,28 @@ func filterNetworks(nws []types.NetworkResource, filter filters.Args) ([]types.N
+			switch fval {
+			case "true", "1":
+				danglingOnly = true
+			case "0", "false":
```
Wonder if we should be explicit here, and add `danglingOnly = false`. Perhaps because I'm used to "fall-through" being the default in other languages, but I had to look twice to interpret this correctly 😊

cpuguy83 replied:
It's already false?

thaJeztah replied:
Yeah, I know, it's probably just me, reading;

```
switch {
  case foo:
    dofoo()
  case bar:
    dobar()
  case baz:
  default:
    dosomethingelse()
}
```

I always tend to first read that as `baz:` falls through to `default:` (so `error` in this case)

thaJeztah replied:
Actually; should this use `filter.ExactMatch()` here? https://github.com/moby/moby/blob/2a7388a6c47a58c402acfe0059fbe27a7a090c23/daemon/prune.go#L185-L191

cpuguy83 replied:
I'm not sure why one would be better than the other. This filters package is pretty confusing.

thaJeztah commented on api/server/router/network/network_routes.go:
```diff
@@ -70,6 +71,8 @@ SKIP:
 		// info is available for network specific GET API (equivalent to inspect)
 		if versions.LessThan(httputils.VersionFromContext(ctx), "1.28") {
 			nr = n.buildDetailedNetworkResources(nw, false)
+		} else if netFilters.Include("dangling") {
```
`Include()` is deprecated; can you change to `Contains()`?

thaJeztah commented on api/server/router/network/network_routes.go:
```diff
@@ -70,6 +71,8 @@ SKIP:
 		if versions.LessThan(httputils.VersionFromContext(ctx), "1.28") {
 			nr = n.buildDetailedNetworkResources(nw, false)
+		} else if netFilters.Include("dangling") {
+			nr = n.buildDetailedNetworkResources(nw, true)
```
Is it a concern that when filtering for "dangling", the returned results will be different than without? (i.e. it having "detailed" information)?

Looking at `buildDetailedNetworkResources()`, and how we use that information during filtering (we're only interested in knowing _if_ containers or services are attached); https://github.com/moby/moby/blob/be146652108f7145e902c275807f9ef71464b031/api/server/router/network/network_routes.go#L359-L403


Wondering if we should have a simplified version, or at least reset the `Containers` and `Services` information afterwards, so that the information that's returned for the network is consistent.

thaJeztah reviewed (changes requested):
Thanks! Left some comments

cpuguy83 commented:
This network code is awful :( We should not be dealing with libnetwork types in the API layer.

thaJeztah commented on api/server/router/network/network_routes.go:
```diff
@@ -70,6 +71,8 @@ SKIP:
 		if versions.LessThan(httputils.VersionFromContext(ctx), "1.28") {
 			nr = n.buildDetailedNetworkResources(nw, false)
+		} else if netFilters.Contains("dangling") {
+			nr = n.buildDetailedNetworkResources(nw, true)
```
Oh, we were discussing this; i.e. can we find a solution where we would still return the same information (i.e., no "detailed" information)? https://github.com/moby/moby/pull/31551#discussion_r162113560

KarthikNayak commented:
@cpuguy83 Hey! are you working on the changes for this? I'm a little busy with something else ATM. Will have some spare time towards the end of the month perhaps.

cpuguy83 commented:
I'm not at the moment.

olljanat commented:
@KarthikNayak this one needs to be rebased

olljanat commented:
Derek add label: status/failing-ci

KarthikNayak commented:
Hey! Sorry for disappearing, will try and pick this up now.

KarthikNayak committed:
Network: add support for 'dangling' filter

Like its counterpart in images and volumes, introduce the dangling
filter while listing networks. When the filter value is set to true,
only networks which aren't attached to containers and aren't builtin
networks are shown. When set to false, all builtin networks and
networks which are attached to containers are shown.

Signed-off-by: Karthik Nayak <Karthik.188@gmail.com>

KarthikNayak commented:
Can someone guide me about why the Windows build is failing?

olljanat commented:
I found couple of other PRs too where TestStartReturnCorrectExitCode has failed and also did try to run it 100 times on loop which this PR passed just fine.

@thaJeztah @cpuguy83 can we get this one merged before its second anniversary? :open_mouth:

cpuguy83 commented on daemon/network/filter_test.go:
```diff
@@ -131,6 +137,16 @@ func TestFilterNetworks(t *testing.T) {
 			name:        "global scope filters",
 		},
+		{
+			filter:      trueDanglingFilters,
```
Sorry to do this, but do you think we can alter the test to actually test for which networks should be returned? Can you also add a network that is not a built-in but is not dangling?

KarthikNayak replied:
@cpuguy83 Haha, I'll get to it soon :+1:

KarthikNayak replied:
@cpuguy83 Done! :+1:

cpuguy83 reviewed (approved):
LGTM

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
whoops; looks like this needs ` gofmt`;

```
12:25:41 daemon/network/filter_test.go:1::warning: file is not gofmted with -s (gofmt)
12:25:42 Build step 'Execute shell' marked build as failure
```

@KarthikNayak can you run `gofmt` on that file to make CI pass?

KarthikNayak commented:
> whoops; looks like this needs ` gofmt`;
> 
> ```
> 12:25:41 daemon/network/filter_test.go:1::warning: file is not gofmted with -s (gofmt)
> 12:25:42 Build step 'Execute shell' marked build as failure
> ```
> 
> @KarthikNayak can you run `gofmt` on that file to make CI pass?

Definitely! Fixed and pushed :)

albers commented:
Thanks for implementing this feature @KarthikNayak.

KarthikNayak commented:
@albers thanks for finishing it up with the bash completion.
